zulipterminal/ui_tools/boxes.py

```diff
@@ -82,19 +84,35 @@ def stream_box_view(self, button: Any=None, caption: str='',
 
     def keypress(self, size: Tuple[int, int], key: str) -> str:
         if is_command_key('SEND_MESSAGE', key):
-            if not self.to_write_box:
-                success = self.model.send_stream_message(
-                    stream=self.stream_write_box.edit_text,
-                    topic=self.title_write_box.edit_text,
-                    content=self.msg_write_box.edit_text
-                )
+            if self.is_edit_mode:
+                if not self.to_write_box:
+                    success = self.model.update_stream_message(
+                        topic=self.title_write_box.edit_text,
+                        content=self.msg_write_box.edit_text,
+                        msg_id=self.msg_id,
+                    )
+                else:
+                    success = self.model.update_private_message(
+                        content=self.msg_write_box.edit_text,
+                        msg_id=self.msg_id,
+                    )
             else:
-                success = self.model.send_private_message(
-                    recipients=self.to_write_box.edit_text,
-                    content=self.msg_write_box.edit_text
-                )
+                if not self.to_write_box:
+                    success = self.model.send_stream_message(
+                        stream=self.stream_write_box.edit_text,
+                        topic=self.title_write_box.edit_text,
+                        content=self.msg_write_box.edit_text
+                    )
+                else:
+                    success = self.model.send_private_message(
+                        recipients=self.to_write_box.edit_text,
+                        content=self.msg_write_box.edit_text
+                    )
             if success:
                 self.msg_write_box.edit_text = ''
+                if self.is_edit_mode:
+                    self.is_edit_mode = False
```

Comment:
You're not resetting `msg_id` here? (though see other comment re combining them)